lex_test.go

```diff
@@ -0,0 +1,314 @@
+package mf2
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_lex(t *testing.T) {
+	t.Parallel()
+
+	var (
+		tokenEOF             = mk(itemEOF, "")
+		tokenExpressionOpen  = mk(itemExpressionOpen, "{")
+		tokenExpressionClose = mk(itemExpressionClose, "}")
+	)
+
+	for _, test := range []struct {
+		name     string
+		input    string // MessageFormat2 formatted string
+		expected []item
+	}{
+		{
+			name:     "empty simple message",
+			input:    "",
+			expected: []item{tokenEOF},
+		},
+		{
+			name:  "text",
+			input: `escaped text: \\ \} \{`,
+			expected: []item{
+				mk(itemText, `escaped text: \ } {`),
+				tokenEOF,
+			},
+		},
+		{
+			name:  "function",
+			input: "{:rand}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemFunction, "rand"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+
+		{
+			name:  "opening function",
+			input: "{+button}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemOpeningFunction, "button"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "closing function",
+			input: "{-button}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemClosingFunction, "button"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "opening and closing functions",
+			input: "{+button}Submit{-button}",
+			expected: []item{
+				tokenExpressionOpen,
+				mk(itemOpeningFunction, "button"),
+				tokenExpressionClose,
+				mk(itemText, "Submit"),
+				tokenExpressionOpen,
+				mk(itemClosingFunction, "button"),
+				tokenExpressionClose,
+				tokenEOF,
+			},
+		},
+		{
+			name:  "variable",
+			input: "{$count :math:round}",
```

Comment:
is ':' character permitted in the standalone function name?

Reply:
yes, MF2 support namespaces in identifiers now - https://github.com/unicode-org/message-format-wg/blob/main/spec/message.abnf#L86